README.md

```diff
@@ -1,7 +1,154 @@
+<!---
+-- Copyright Kani Contributors
+-- SPDX-License-Identifier: Apache-2.0 OR MIT
+---> 
+
 ## rust-lean-models
 
 Lean models of various Rust libraries to facilitate Lean-based verification of Rust programs.
 
+
+## Description
+The library defines equivalent Lean versions of functions from the Rust standard library.
+This library is intended to be used for verifying string-manipulating Rust programs via Lean.
+
+The Lean implementation is based on the description of the Rust functions, which are published at https://doc.rust-lang.org/std.
+The library includes:
+- Definitions of functions equivalent to those from the Rust standard library
+- Proofs that these definitions are correct (i.e., match our understanding of the Rust original)
+- Supporting lemmas and theorems for proof construction
+
+## Installation
+
+To use the library, add the following lines to your `lakefile.lean`:
+
+` require rust-lean-models from git "https://github.com/model-checking/rust-lean-models" `
+
+Then do `lake update` in the terminal.
+
+## Usage
+### Translating Rust program 
+
+For any Rust built-in functions which are used in user code , map it with 
+the corresponding function name in the library (see the Tables below).
+
+### Proof Tutorial
+We demonstrate some proof techniques in Proof_Tutorial.lean through two simple programs that compute
+the longest common prefix and longest common substring of the two input strings. 
+More examples can be found in Proof_Example.lean
+
+## Implementation
+
+
+### Recursive function definition
+For each Rust function, we provide a recursive Lean function. Implementing 
+the equivalent functions in Lean recursively enables user to construct 
+induction proofs conveniently. The Lean function has the same as the Rust original, 
+except when the Rust name  clashes with a Lean keyword. In case of a clash, a Rust function 'func_name' 
+is renamed to `rust_func_name` in Lean.
+
+Implementing a function recursively may requires some extra parameters.
+In those cases, first, we implement an auxiliary function (name: `func_name_aux`) which is defined 
+recursively with the parameters, then we define the function `func_name` based on the auxiliary function 
+with initial values for the parameters. 
+For example, `char_indices` is defined based on `char_indices_aux` as 
+`def char_indices (s: Str):= char_indices_aux s 0`.
```

Comment:
```suggestion
`def char_indices (s: Str) := char_indices_aux s 0`.
```

Reply:
Can you add "where the parameter is ..." and describe what the parameter denotes.